SQL-like column return for non-equi and rolling joins

sritchie73 opened the pull request:
#1615
#1700 
#2006 
#2569 
Other related issues already closed (e.g. dups) : #1469, #1807, #2307, #2595, #2602 

Here's a diagram illustrating the issue for a simple non-equi join:
```
dt_left[dt_right, on = .(by_left > by_right)]
```
![screen shot 2018-03-26 at 2 20 06 pm](https://user-images.githubusercontent.com/1674387/37885440-ee495962-3100-11e8-985b-6da09a5b266f.png)

When performing any join, only one copy of each key column is returned in the result. Currently, the column from `i` is returned, and labelled with the column name from `x`, making equi joins consistent with the behaviour of base `merge()`.

This causes confusion for many people (e.g. #1700 and referring issues). I'm working on the DataCamp course on joins with `data.table`, and would like to avoid teaching the often recommend but inelegant approach of creating copies for the join, e.g.:

```
dt_left[, copy.x := by_left]
dt_right[, copy.y := by_right]

joined <- dt_left[dt_right, on = .(copy.x > copy.y)]
joined[, copy.x := NULL]

dt_left[, copy.x := NULL]
dt_right[, copy.y := NULL]
```

I've consulted with @arunsrinivasan, who has indicated the preferred behaviour is essentially what the code above achieves -- keep the columns from both `x` and `i` if they are used in a non-equi (or rolling) join. This is also consistent with the behaviour of SQL.

This PR changes the behaviour of non-equi and rolling joins to this end, e.g.:

```
dt_left[dt_right, on = .(by_left > by_right)]
```
![screen shot 2018-03-28 at 11 03 49 am](https://user-images.githubusercontent.com/1674387/38001293-c2a6b02a-3277-11e8-9e73-b3f4e8795e1b.png)

TODO:

 - [x] Extend the patch to work for cases where there arent any equi join keys.
 - [x] Extend patch to work under `!missing(j)`
 - [x] Extend patch to work under `by = .EACHI` (`byjoin`) 
 - [x] Find and link other related issues to this PR
 - [x] Update unit tests for non-equi and rolling joins that expected 1 returned column per key
 - [x] Identify any failing unit tests caused by bug regression rather than change in expected output
 - [x] Fix any bug regressions 
 - [x] Check `test.data.table(with.other.packages = TRUE)`
 - [x] Create any new necessary unit tests
 - [x] Update any related documentation

sritchie73 commented:
I've updated the following unit tests since they were not expecting columns from both `x` and `i` to be returned (also detailed in the git commit message). These include those covering functionality extending packages in Suggests:

- 129: added additional V1 column from J() to expected output.
- 130: added additional V1 column from J() to expected output.
- 131: added additional V1 column from J() to expected output.
- 148: added additional V1 column from CJ() to expected output. 
- 149: added additional V1 column from J() to expected output.
- 917: updated expected output. added i.ts and updated ts (from x) to contain matched values from x$ts instead of matched values from i$ts.
- 1317.1: updated expected output. added i.y and updated y (from x) to contain matched values from x$y instead of matched values from i$y.
- 1317.2: updated expected output. added i.y and updated y (from x) to contain matched values from x$y instead of matched values from i$y.
- 1470.1: rolling join on J(), renamed x column from output as V1 and put in correct position, then added x as DT$x. 
- 1470.2: rolling join on J(), renamed x column from output as V1 and put in correct position, then added x as DT$x.
- 1470.3: rolling join on SJ(), renamed x column from output as V1 and put in correct position, then added x as DT$x.
- 1614.1: updated expected output. renamed x column from output as V1 and put in correct position, then added dt$x column to expected output.
- 1614.2: updated expected output. renamed x column from output as V1 and put in correct position, then added dt$x column to expected output.
- 1614.3: updated expected output. renamed x column from output as V1 and put in correct position, then added dt$x column to expected output.
- 1614.4: updated expected output. renamed x column from output as V1 and put in correct position, then added dt$x column to expected output.
- 1660.1: updated expected return columns.
- 1660.2: updated expected return columns.
- 1682.1: Updated `by` statement in chained operator to use column names in dt2 to match columns used by `by=.EACHI` 
- 1682.2: Updated `by` statement in chained operator to use column names in dt2 to match columns used by `by=.EACHI` 
- 1682.3: Updated `by` statement in chained operator to use column names in dt2 to match columns used by `by=.EACHI` 
- 1682.4: Updated `by` statement in chained operator to use column names in dt2 to match columns used by `by=.EACHI` 
- 1682.5: Updated `by` statement in chained operator to use column names in dt2 to match columns used by `by=.EACHI` 
- 1682.6: Updated `by` statement in chained operator to use column names in dt2 to match columns used by `by=.EACHI` 
- 1682.7: Updated `by` statement in chained operator to use column names in dt2 to match columns used by `by=.EACHI`

sritchie73 commented:
The following unit tests fail due to bug regressions, or new bugs (test output differs from expected new behaviour):

- 1469.2: after roll DT now has a key when it shouldnt
- 1660.1: "End" column missing from output, and Location column has contents of dt1$end even though it is involved in a non-equi join.
- 1660.2: "Location" column missing from output.
- 1682.1: Column V2 is missing from non-equi join result 
- 1682.2: Column V2 is missing from non-equi join result
- 1682.3: Column V2 is missing from non-equi join result
- 1682.4: Column V2 is missing from non-equi join result
- 1682.5: Column V2 is missing from non-equi join result
- 1682.6: Column V2 is missing from non-equi join result
- 1682.7: Column V2 is missing from non-equi join result

I also need to follow up on the following tests which fail, but may not be bugs:

- 1540.15: unexpected difference in column order (but not contents) after rolling join.
- 1614.4: values in dt$x do not match those expected based on values in dt$y, but this might be due to rollends=TRUE. 
- 1846.1: columns do not match - check that handling of 0-row dt in i is correct based on linked issue.
- 1846.2 columns do not match - check that handling of 0-row dt in i is correct based on linked issue.
-

sritchie73 commented:
This PR is ready to review. 

Many unit tests had to be updated to reflect the new column return. I've listed the changes made above. I've tested with all suggested packages on OSX, but without OpenMP support (as far as I can tell the only tests covering this are `1760.x` which tests `fread()`). I've also tested `with.other.packages = TRUE`. @jangorecki these changes break `dtq::dtl()`, tests `13.2` and `13.3` fail in other.Rraw (the output of `dtl()` is empty) - but all data.table queries in the setup return identical results to the master branch.

In addition to the obvious change of returning columns from both `x` and `i`, there are a few other side effects to consider which may or may not break peoples existing code:

 - When joining on a constructed `data.table` (e.g. through `J()`, `SJ()`, `CJ()`) the result will also contain any columns constructed in these operators as `V1` ... `VN`
 -  When joining on an empty `data.table` columns from that `data.table` will be returned, filled with `NA` values.
 - In existing code, any references to the column previously returned by a non-equi or rolling join will now contain different values, since that reference will now extract the column from `x` where they previously would contain values from that column in `i`.
 - When chaining operators, to achieve the same result as `by = .EACHI` using `by` in the second operator you now have to specify the keys from both `x` and `i`. E.g. `x[i, on=.(a>b), .N, by=.EACHI]` is now equivalent to `x[i, on=.(a>b)][,.N, by=.(a=b)]`.

sritchie73 commented:
Thanks @jangorecki - I don't have any further work planned on this PR. Unfortunately its not possible to split apart the PR – all changes in `data.table.R` are needed, otherwise `tests.Rraw` will crash outright (rather than tests failing). 

This now needs Matt's feedback.

The changes in `data.table.R` can be summarised as follows:

 1. Lines 595–614: drop any non-equi or rolling join column indices from `leftcols` and `rightcols` after `bmerge()`, so that they are kept from both `x` and `i` by the processing of `j`.
 2. Lines 758–767: additional `if` statement in the processing of `j` that handles the special case where `leftcols` and `rightcols` become empty in step 1 (they are set to `NULL` to because the empty vector is used for other cases).
 3. Lines 1232–1242: restore `leftcols` and `rightcols` so that they again contain non-equi or rolling join columns, so that grouping of rows in the `by` processing is correct (i.e. performed the same way as in the master branch).

This is the minimum set of changes required for `tests.Rraw` to run without crashing entirely.

With those changes made, I've been through each failing unit test to determine whether they fail because (1) the expected columns needed to be updated, (2) the changes introduced a new bug, or (3) a bug regression. I've only changed unit tests to fix case (1): so that they expect columns from both `x` and `i` - and also made sure that this doesn't invalidate the test in other ways (i.e. they should still test the bug/feature they were originally designed to catch). More eyes are definitely needed on these.

mattdowle commented:
I've done a first sweep through the linked issues and the changed tests.  Thanks for taking this on!  First impression is that many people have been similarly confused and that this PR should go ahead.  As Jan suggests, controlled by an option please, and a new item 3 in the top section of news linked to all the issues that will be solved.  For all the tests that have changed, their old test could be copied into a new section where they are run with the option off, to ensure the option works.

sritchie73 commented:
Thanks @mattdowle ,

I've added an option, `bothjoincols`, which when `TRUE` (default) will return non-equi join and rolling join columns from both x and i. 

In tests.Rraw i've reverted all the above tests to their code in master, adding `bothjoincols=FALSE` so they work as previously intended. I've copied their updated code from commit 9d6c44d to the end of tests.Rraw (tests 1899.01 – 1899.27) with comments linking back to their original test numbers.

Before I write the news entry:

- Is there a better name for the argument?
- Do you want the default to be FALSE for release 1.10.5?
- Do you want an option in `options()` controlling the default value?

jangorecki commented:
For me definitely default FALSE, option.

sritchie73 commented:
Thanks for your feedback and patience @jangorecki - I've added an option and set to FALSE - this seems like the right approach given the other NEWS entries on breaking changes.

mattdowle commented on R/data.table.R:
```diff
@@ -1202,6 +1231,17 @@ chmatch2 <- function(x, table, nomatch=NA_integer_) {
+        ivars = names(i)
+        w2 = chmatch(ansvars[wna], ivars)
+        if (any(w2na <- is.na(w2))) {
+          ivars = paste("i.",ivars,sep="")
```
According to codecov, these two lines are missing coverage causing diff coverage to be 94.28%. That's pretty good to be so few! Is there an easy way to cover them please? Diff coverage needs to be 100% unless there's a good reason; e.g. unusual stop()'s that are difficult to simulate to trigger.

MichaelChirico commented:
check the latest master, I just purged (or tried at least) all usage of
paste(sep="")

On Tue, Apr 10, 2018, 4:19 PM Scott Ritchie <notifications@github.com>
wrote:

> *@sritchie73* commented on this pull request.
> ------------------------------
>
> In R/data.table.R
> <https://github.com/Rdatatable/data.table/pull/2706#discussion_r180332234>
> :
>
> > @@ -1202,6 +1231,17 @@ chmatch2 <- function(x, table, nomatch=NA_integer_) {
>          xcols = w
>          xcolsAns = seq_along(ansvars)
>          icols = icolsAns = integer()
> +      } else if (is.null(leftcols)) {
> +        xcols = w[!wna]
> +        xcolsAns = which(!wna)
> +        ivars = names(i)
> +        w2 = chmatch(ansvars[wna], ivars)
> +        if (any(w2na <- is.na(w2))) {
> +          ivars = paste("i.",ivars,sep="")
>
> I've added a new test that will cover those lines.
>
> @jangorecki <https://github.com/jangorecki> - the code on lines 1239 –
> 1242 (including paste(... sep="")) are written to match the code on lines
> 1256 – 1260.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Rdatatable/data.table/pull/2706#discussion_r180332234>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHQQdU9y6tgoAli1n2jaYnPRJpx-y_XGks5tnGsNgaJpZM4S6eaW>
> .
>

MarkusBonsch commented:
I love this. It will avoid so much confusion for new users of `data.table`. But it will also cause a lot of anger from people whose code breaks . . . maybe even silently because a column with the same name may still exist but contain different values as far as I understand the new behaviour. While I would vote for it, it needs to to very carefully decided.

MarkusBonsch commented:
Since this is a breaking change, it is a chance to really consolidate join behaviour thoroughly. 
There are 2 other issues that come to my mind:
1. Weird behaviour can also occur on non-rolling equi joins with `nomatch=NA` as pointed out here: #1615 
   ```
   library(data.table)
   options(datatable.bothkeycols = TRUE)
   x = data.table(a=1:3, w=letters[1:3])
   y = data.table(b=3:5, z=6:4)
   x[y, on=c(a="b")]
   # a    w z
   # 1: 3    c 6
   # 2: 4 <NA> 5
   # 3: 5 <NA> 4
   ```
   Here column a should be `3, NA, NA`. The reason that it is not,  is that `a` is not `x.a` but `i.b`, so this 
   could be fixed by really returning `x.a` in equi joins.

2. The column that is returned from `i` may have a different type than the original column in `i` due to coercion during the join:
   ```
   library(data.table)
   options(datatable.bothkeycols = TRUE)
   dt <- data.table(x = letters[1:10], y = factor(letters[1:10]))
   class(dt$x)
   # "character
   str(dt[dt, on="x>y"])
   # Classes ‘data.table’ and 'data.frame':	46 obs. of  4 variables:
   # $ x  : chr  "b" "c" "d" "e" ...
   # $ y  : Factor w/ 10 levels "a","b","c","d",..: 2 3 4 5 6 7 8 9 10 3 ...
   # $ i.x: Factor w/ 10 levels "a","b","c","d",..: 1 1 1 1 1 1 1 1 1 2 ...
   # $ i.y: Factor w/ 10 levels "a","b","c","d",..: 1 1 1 1 1 1 1 1 1 2 ...
   # - attr(*, ".internal.selfref")=<externalptr>
   ## huups, i.x is returned as a factor
   ```
   This behaviour can IMHO be fixed by making the type conversion local in bmerge. Then it is important 
   to return the column from x in equi joins (similar to point 1 above) so that the type is correct in these 
   cases.

I don't have a very strong opinion on these two points, but since there is a chance to fix them now, we should decide this actively. From the top of my head, I would go for it because I think the change is easy to make and both points can come as a bad surprise to the unaware user.

sritchie73 commented:
I think it would be better to create separate PRs for both those issues.

sritchie73 commented:
Does your PR #2734 fix point 2?

MarkusBonsch commented:
No, #2734 dosn't tackle the issue. It only problems solves in type conversion.
I agree, that the two points should be seperate PRs but they should go live in the same CRAN release and documented together, because they introduce braking changes in the same place as your PR. All that conditional on us wanting to implement them. What do you think about it, would you go for it?

I will create issues for each point and maybe propose a PR if I find the time. Then we can discuss, whether we want to implement them

mattdowle commented:
I like the idea of including both sides of the rolling and non-equi join columns in the result by default when `j` is missing, but that seems separate to the change in value for the existing column name (as @MarkusBonsch referred to above, too).

This is the way I use rolling joins : 
```R
> DT = data.table(id=c("foo","foo","bar"), date=c(1L,3L,5L), val=6:8, key="id,date")
> DT
       id  date   val
   <char> <int> <int>
1:    bar     5     8
2:    foo     1     6
3:    foo     3     7
> options(datatable.bothkeycols = FALSE)
> DT["foo"]
       id  date   val
   <char> <int> <int>
1:    foo     1     6
2:    foo     3     7
> options(datatable.bothkeycols = TRUE)
> DT["foo"]          # no change for equi-join; i.e. good current behaviour
       id  date   val
   <char> <int> <int>
1:    foo     1     6
2:    foo     3     7

> options(datatable.bothkeycols = FALSE)
> DT[.("foo",2)]
       id  date   val
   <char> <int> <int>
1:    foo     2    NA
> options(datatable.bothkeycols = TRUE)
> DT[.("foo",2)]     # no change for equi-join where missing; i.e. good current behaviour
       id  date   val
   <char> <int> <int>
1:    foo     2    NA
```
So far, so good. Notice how the default column names in `i` (i.e. V1, V2) are not significant above and those column names are discarded. Where columns are joined, the data in that column is the same kind of data. Like dates, or ids, or some such.  I like the column name `date` from `x` to be returned because that's what I've joined to and I didn't need to name the `i` columns. I like the behaviour above.

Let's specify `roll=` : 
```R
> options(datatable.bothkeycols = FALSE)
> DT[.("foo",2), roll=TRUE]     # current behaviour
       id  date   val
   <char> <int> <int>
1:    foo     2     6
> options(datatable.bothkeycols = TRUE)
> DT[.("foo",2), roll=TRUE]     # new behaviour
       id  date   val    V2
   <char> <int> <int> <int>
1:    foo     1     6     2
```

Suddenly, just by specifying `roll=`, the data in the `date` column has changed.  (We get an extra column in the result too, but an extra column per se is nice and welcome).  Most often, it's the data in `V2` that I want to use afterwards, though.  The data in V2 is my regular time series that I have rolled from.  I might want to plot that against some observation such as price and see a flat line though periods of no trading, for example.

I think I'd prefer the following behaviour : 
```R
> DT[.("foo",2), roll=TRUE]     # proposed change to PR
       id  date   val x.date
   <char> <int> <int>  <int>
1:    foo     2     6      1
```
So, the data in the join columns from `i` is always and consistently returned with the names from `x` (as is true for equi joins too).  And the prefix `x.` is always placed on the extra columns (with the same column name from `x`).  Some advantages of this are :  
1. the `i.` prefix on the extra columns is not present or not depending on whether the join column names are the same or not (which is currently the case in the PR).  Generally, we try and avoid changes in result that depend on data.
2. there is no change to existing production code using existing column names of results.  New code can use the extra columns if and when it wishes to.
3. it's still more helpful to folk new to data.table. With the option turned on (new behaviour) they will see two columns returned and by looking at the data they can see what's happened.

Note that `x.date` is already usable in `j` before this PR and already does what I propose. So the proposal is really just a small tilt to add that in to the result rather than change existing column name. Here's some examples using named columns in `j` : 

```R
> options(datatable.bothkeycols = FALSE)
> DT[.("foo",2), .(id, date, i.V2, x.date), roll=TRUE]
       id  date  i.V2 x.date        # existing behaviour
   <char> <int> <int>  <int>
1:    foo     2     2      1
> options(datatable.bothkeycols = TRUE)
> DT[.("foo",2), .(id, date, i.V2, x.date), roll=TRUE]
       id  date  i.V2 x.date        # under this PR with option on
   <char> <int> <int>  <int>
1:    foo     1     2      1
> DT[.("foo",2), .(id, date, i.V2, x.date), roll=TRUE]
       id  date  i.V2 x.date        # proposed change to this PR back to existing behaviour
   <char> <int> <int>  <int>
1:    foo     2     2      1
```

It will be an easier sell if there is no change to any query that includes a `j=` clause.  The only change would be if `j=` is missing.  If that's possible to achieve whilst still solving all the complaints, then that would be better.  That way, production code using `roll=` or `on=` need only be checked where `j=` is missing and there could be an option provided to print all such calls to aid migration.

I've focused on rolling joins here. I haven't considered non-equi joins yet. But thoughts so far?

MarkusBonsch commented:
I agree with @mattdowle's proposal for the rolling joins. His examples convince me that the current behaviour makes sense for rolling joins.

However, from my perspective, the issue is completely different for non-equi joins.

Example: imagine, we have a dataset of trees with type, height, and diameter:
```
library(data.table)
trees <- data.table(type = c("oak", "oak", "birch", "birch", "maple", "maple"),
                    height = c(20, 45, 15, 33,  45, 10),
                    diameter = c(1.5, 0.9, 2.4, 0.4, 0.7, 0.8))
```

Now imagine, we have a specification for trees that are needed for a specific site with type and minimum height:
```
requirements <- data.table(type = c("oak", "birch", "maple"),
                           minimumHeight = c(30, 10, 55))
```
Now, we are using non-equi join to see, which of our trees are good for the requirements:
```
options(datatable.bothkeycols = FALSE)
trees[requirements, on=c("type==type", "height>=minimumHeight")]
#    type height diameter
#    1:   oak     30      0.9
#    2: birch     10      2.4
#    3: birch     10      0.4
#    4: maple   55       NA
```
This result is misleading in two regards:
1. It suggests, we have a maple tree with height 55 and unknown diameter that matches the requirements while in reality, we have no maple tree that matches the requirements.
2. It suggests that our two birch trees that match the requirements both have a height of 10 and the oak has a height of 30, while in realty, they all have different heights. If we want to choose the shortest possible tree from the matching ones, we can't do this from the result

The new behavior here is much more informative:
```
```
options(datatable.bothkeycols = TRUE)
trees[requirements, on=c("type==type", "height>=minimumHeight")]
#    type height diameter
#    1:   oak     45      0.9
#    2: birch     15      2.4
#    3: birch     33      0.4
#    4: maple     NA       NA
```
We immediately see
1. We don't have a matching maple tree (OK, we could have achieved that with nomatch = 0L in the old version)
2. We get the correct heights of the other trees and can see that the second birch tree is an overkill for the requirements so we should use the first one.

Therefore, I think that the new behaviour makes a lot of sense for non-equi joins. However, if we want to avoid breking existing code, we could just add the column x.height. Then the user would see the weird behaviour of the height column in the result.

mattdowle commented:
Great example, @MarkusBonsch.  Yes I agree non-equi joins are very different, then.  Am I right in thinking that rolling join and non-equi join are mutually exclusive?  i.e. a rolling join makes no sense at the same time as a non-equi join. Rolling joins join each row in `i` to one row in `x`, one-to-one.  Whereas non-equi join each row in i to usually multiple rows in `x`; e.g. `by=.EACHI` might be appropriate.  Whereas I can't imagine using `by=.EACHI` together with rolling join.   It does make sense to use both `on=` and `roll=` together, but only as follows, how about : 

1.  When `roll=` is specified, change this PR as I proposed.  `on=` may also be specified but only with equi conditions. If any non-equi conditions are present in `on=` then that's an error, as should be the case already.
-vs-
2. When `roll=` is not specified and `on=` contains any non-equi conditions then this PR remains as-is. The columns with non-equi conditions would both be present in the result, with the same names from `x` and `i`. It would be breaking change (different data in the same column name) but there would no longer be any weird behaviour and it's more like a bug fix.  After all, non-equi joins are much more recent than rolling joins, so it's more reasonable that a breaking change be made for them as a result of feedback.

This would not be different behaviour that depends on the _data_, but different behaviour that depends on the _syntax_, which is normal and ok.

mattdowle commented:
@sritchie73  We seems to have agreement. What you think?

sritchie73 commented:
@mattdowle - I'm wondering whether the example you've given isn't so much about rolling joins, but about joins to an anonymous data.table / vector.

When `i` is a data.table stored in a variable, I think it often still makes sense to keep both columns – other resources that teach rolling joins take this approach (e.g. https://r-norberg.blogspot.co.uk/2016/06/understanding-datatable-rolling-joins.html).  

I'm wondering if a better solution is to have `bothkeycols = FALSE` when `i` is an anonymous data.table (created inside `[]` using `.()`, `J()`, `CJ()`, etc), in which case I suspect you implicitly do not care about the columns in that data.table.

MarkusBonsch commented:
Good point @sritchie73, it really seems that on rolling joins with non-anonymous i, the old behavior is confusing. I, therefore, revise my opinion and think that the `bothkeycols` implementation as suggested by you makes sense for rolling joins.
A small example illsutrates, why I have changed my mind.

Assume we have a machine that produces chewing gum. Unfortunately, the machine failed too often in the past:
  ```
library(data.table)
failures <- data.table(failureDate = as.POSIXct(c("2017-01-01", "2017-03-25", "2017-01-02", "2017-06-19", "2018-03-01")))
failures
#   failureDate
# 1:  2017-01-01
# 2:  2017-01-02
# 3:  2017-03-25
# 4:  2017-06-19
# 5:  2018-03-01
```
The temperature is a very crucial operating parameter of the machine and the engineers suspect that wrong temperature settings cause the failures of the machine.
From the machine, we can extract a parameter log with a full history of temperature changes:
```
temperatureChanges <- data.table(changeDate = as.POSIXct(c("2016-12-01", "2017-02-10", "2016-05-02", "2017-01-19", "2017-03-30")),
                             newVal = c(73, 72.5, 74, 69.5, 70))
temperatureChanges
#     changeDate newVal
# 1:  2016-12-01   73.0
# 2:  2017-02-10   72.5
# 3:  2016-05-02   74.0
# 4:  2017-01-19   69.5
# 5:  2017-03-30   70.0
```
Now, we want to know, which temperature was set at each failure - a classical task for a rolling join. In master it looks like this:
```
options(datatable.bothkeycols = FALSE)
temperatureChanges[failures, on = "changeDate==failureDate", roll = TRUE]
#   changeDate newVal
# 1: 2017-01-01   73.0
# 2: 2017-01-02   73.0
# 3: 2017-03-25   72.5
# 4: 2017-06-19   70.0
# 5: 2018-03-01   70.0
```
There are two issues with this:
1. The column `changeDate` doesn't actually contain the dates where temperature was changed, but the `failureDates` of the machine!
2. (very minor) we don't see when the temperature values were actually changed.

Now in the new implementation:
```
options(datatable.bothkeycols = TRUE)
temperatureChanges[failures, on = "changeDate==failureDate", roll = TRUE]
#   changeDate newVal failureDate
# 1: 2016-12-01   73.0  2017-01-01
# 2: 2016-12-01   73.0  2017-01-02
# 3: 2017-02-10   72.5  2017-03-25
# 4: 2017-03-30   70.0  2017-06-19
# 5: 2017-03-30   70.0  2018-03-01
```
From columns 2 and 3, we can immediately see the temperature setting at each failure with proper column names. Column 1 additionally provides the change date of the temperature.
From my perspective a much more convenient and reliable output.

sritchie73 commented:
I'll propose the following in addition to the changes already in this PR:

1. Handling of anonymous rolling join columns

Following the use case given by @mattdowle, when `options(datatable.bothkeycols = TRUE)` and a rolling join is performed on an anonymous `data.table` then the rolling join column will take the name of the column in `x`, and the column in `x` will have the prefix `x.`

2. General handling of anonymous join columns

More generally, when `options(datatable.bothkeycols = TRUE)`, it might be useful to adopt the above behaviour for all anonymous join columns (e.g. both rolling join and non-equi join columns). If a column is used multiple times in on (e.g. `on = .(date >= V1, date <= V2)`) The output columns could be numbered, e.g. (`date.1` for `V1`, `date.2` for `V2`, and `x.date` for `date`). This could be controlled through another option, e.g. `options(datatable.renameanonymousjoincols)` (this option would also control the behaviour in the case of rolling joins as in 1. above).

3. Values in join key columns.

When `options(datatable.bothkeycols = FALSE)` and for equi-joins, the values in the key column should reflect the values in the column from `x` rather than from the column from `i`. In addition to solving the above scenario(s), this will also fix the unexpected behaviour when `nomatch = NA` that @MarkusBonsch pointed out. This would controlled by another option, e.g. `options(datatable.xvaluesinjoinkey)`.

mattdowle commented:
@sritchie73  I'm not sure you fully understood my comment of 28 days ago above [here](https://github.com/Rdatatable/data.table/pull/2706#issuecomment-380594104). Can we revisit that before continuing.  I was agreeing to keeping both columns in a rolling join, but did you think I was not, since you wrote :  
> When i is a data.table stored in a variable, I think it often still makes sense to keep both columns – other resources that teach rolling joins take this approach (e.g. https://r-norberg.blogspot.co.uk/2016/06/understanding-datatable-rolling-joins.html).

Where I wrote "# proposed change to this PR back to existing behaviour" I was referring just to when the columns in `j=` are specified as was the context of that comment (and that is to do with naming).  When `j=` is not provided, then I showed in that comment an extra column should be included in the result for the rolling join column (in agreement with the article you linked to).  I see in my comment that my 4th bullet point was off track because there would not be "one or more", but exactly one always: because the columns before the last joined column are equi-joined and nobody is suggesting keeping both sides of those iiuc (I don't see `name` column duplicated in the linked article for example, just both sides of the rolled column).  Then @MarkusBonsch  came in and changed his mind and gave an example of keeping both sides of the rolling join column. This seems to have stemmed from you saying that I was suggesting not included both sides of the rolling join column!  Can you both read my earlier and longer comment above please.
 
If we are in fact in agreement about the presence and contents of columns in the result, are we now onto deciding the naming of them?   In an n-column rolling join, the first n-1 columns are equi joined. Only the nth column is rolled.  The equi-join columns of a rolling join are like a regular equi-join, not duplicated in the result, and receive their names from `x` regardless of whether `i` column names are anonymous or not. Right?  Lets get straight on that terminology and then we should be there.

Isn't point 1 in @sritchie73's last comment exactly what I already suggested (extra column in the result having `x.` prefix).  But this is what @MarkusBonsch's example (and the article attached) show is nicer to use the name from `i` for the rolling column.

ethanbsmith commented:
While this PR is focused on non-equi-joins, it may be worth reconsidering the behavior of equi-joins, as that may be the source of the problems. lets take the example from  @mattdowle comment from Apr 11:

```
> DT = data.table(id=c("foo","foo","bar"), date=c(1L,3L,5L), val=6:8, key="id,date")
> DT
       id  date   val
1:    bar     5     8
2:    foo     1     6
3:    foo     3     7

> DT[.("foo",2)]
       id  date   val
1:    foo     2    NA
```
Now the above is interesting because it looks very similar to an indexed lookup `DT[id == "foo"]` however  that is not what is going on. The output, documentation and many discussions are all very clear that this is an outer join. The difference is expressed below:
```
> DT[id == "baz" & date == 2]
Empty data.table (0 rows) of 3 cols: id,date,val

> DT[.("baz", 2)]
    id date val
1: baz   NA  NA
```
As we see, the indexed lookup returns and empty table of the same shape as `DT`, while the outer join (as clearly explained in @sritchie73 original comment) returns the expected number of rows, but the columns start to be questionable. Lets look at a further example:
```
> DT[.("baz",2,"chocolate", "cake")]
    id date val        V3   V4
1: baz    2  NA chocolate cake
```
what we begin to see here is the true nature of the outer join. what we expect is all rows from `i` and matching rows from `x`, with all columns from both `x` and `i`. However, what starts to come into focus is that we're not really getting all columns from both `x` and `i`. The default column names fro `V3` and `V4` make sense, but `V1` and `V2` are missing.
What I suspect is goin on here is that this is being treated as shorthand for something liek this:
```
> DT[data.table(id="baz", date= 2, "chocolate", "cake", key = c("id","date"))]
    id date val        V3   V4
1: baz    2  NA chocolate cake
```
i.e. an outer join between 2 tables based on their keys. This is a very nice shorthand and is suspect works in many use cases. however what if we want this:
```
> DT[data.table(id = "baz", missing.date = 5, "chocolate", "cake", key = c("id", "missing.date"))]
    id date val        V3   V4
1: baz    5  NA chocolate cake
```
well, now we see that the shorthand is overiding our column names on the outer join and prduces misleading results, even though we explicitly spcified column names. Since DT has no rows with `date == 5`, how are we getting output with such a result? Even further, why does the code below even work?
```
> DT[data.table(oldid = "baz", missing.date = 5, "chocolate", "cake")]
    id date val        V3   V4
1: baz    5  NA chocolate cake
```
In the code above, the anonymous table does not have a key, nor does it have any common column names. why is the auto-join being done? If I explicitly set the key on the `i` table, the query fails as expected:
```
> DT[data.table(id = "baz", missing.date = 5, V3 = "chocolate", V4 = "cake", key = c("V3,V4"))]
Error in bmerge(i, x, leftcols, rightcols, io, xo, roll, rollends, nomatch,  : 
  typeof x.date (double) != typeof i.V4 (character)
```
Now lets consider one last very interesting scenario with new data:
```
> DT2 = data.table(id = c("foo", "foo", "bar", "baz"), date = c(1L, 3, 5L, NA), val = c(6:8, NA), key = "id,date")
> DT2
    id date val
1: bar    5   8
2: baz   NA  NA
3: foo    1   6
4: foo    3   7

> DT2["baz"]
    id date val
1: baz   NA  NA

> DT["baz"]
    id date val
1: baz   NA  NA
```
how do we distinguish this new scenario with `DT2` from the scenario with the original `DT`? the `DT2` query matched a row, while the `DT` query did not and there is no way to tell the difference from the output.

I suggest that the heart of the problem is that both the processing and the output format has been optimized for the shorthand notation that assumes the keys from both tables have the same column names, but is failing to produce expected processing when they don't. Even on equi-joins. 
I accept that this shorthand is very useful and quite like a very common use case, but it's hard to generalize and extend to other scenarios. I suspect that if you start with the assumption that current `data.table` equi-join processing is good and valid, then it will be impossible to come up with a generalized solution output formats for all joins

Might I suggest that you consider:L@sritchie73
1. formalize the treatment of anonymous non-keyed tables in `i` as if they have the same column names and key as `x` and document this
2. Apply @sritchie73 original solution to ALL joins

I understand that this would produce breaking changes in exiting code on equi-joins but they have the following advantage:
1. consistent output across all types of joins in all scenarios. simply changing a `==` to `>=` or vice-versa does not wreak havoc on output shape or column names
2. joins work like SQL joins and no more hurt brains 
3. fixes the problem of masked column names that drove many of the original issue reports

MarkusBonsch commented:
@mattdowle I have revisited your post and see your point. I think that my understanding is in agreement with your proposal from 28 days ago.
In my own words, I would prefer to see the following behavior for rolling joins:
* Keep both columns, the one from `x` and the one from `i`
* if they have different names, they will retain their original names (as in my example above failureDate and changeDate)
* if they have the same name, e.g. `testCol`, then the column `testCol` in the result will be the column from `i` and the column from `x` will be retained in the result as `x.testCol`.

MarkusBonsch commented:
@ethanbsmith I don't want to commetn on your proposal yet, I just observed some things.
You wrote:
>
 ```
DT[.("baz", 2)]
    id date val
1: baz   NA  NA
```
On my machine with the current master, the result is:
```
DT[.("baz", 2)]
    id date val
1: baz   2  NA
```
which is less suprising to me after your first example.

You also wrote:
> Even further, why does the code below even work?
`DT[data.table(oldid = "baz", missing.date = 5, "chocolate", "cake")]`

IIUC, the first two columns of `i`, `oldid`, `missing.date` are automatically interpreted as the join columns. So reproducing this with a keyed example would mean:
```
DT[data.table(oldid = "baz", missing.date = 5, V3 = "chocolate", V4 = "cake", key = c("oldid","missing.date"))]
```
which yields the expected result
```
    id date val        V3   V4
1: baz    5  NA chocolate cake
```
Your example with the key on `V1` and `V2` attempts to do something different and fails correctly because of data types.

ethanbsmith commented:
@MarkusBonsch 

1. you're right, thanks for catching this. bad copy/paste . i don't think this affects the line of thought though. I now will recheck all results and edit post and annotate edits where needed. 

2. again, you're right. this certainly did not make the point i was going after, which was that behavior is different when keys are explicit vs implied. i think the line of thinking holds up without this example. I don't want to distort the record, so i will leave it in place, but I think we can ignore this example from the discussion.

mattdowle commented:
@ethanbsmith Thanks for the detailed comments and thoughts.  Have read it once in detail and will start to respond.  Just before I do,  have you seen [FAQ 2.14](https://cran.r-project.org/web/packages/data.table/vignettes/datatable-faq.html#can-you-explain-further-why-data.table-is-inspired-by-ab-syntax-in-base)?  There I explain what inspired `X[Y]` syntax. When I see `X[Y]` I think to myself "looking up";  rows in X are being looked up using Y,  like a matrix subset.  I still like this thinking. I don't think join or outer join, and I'm freed from having to remember which way around "left" and "right" is,  since `Y` is inside the bracket in `X[Y]` so it's `X` being looked up.

You wrote : 
> In the code above, the anonymous table does not have a key, nor does it have any common column names. data.table joins are done on keyed tables without specifying an on parameter, but why is the auto-join being done with no explicit key? It seems that the implied key processing is based on column data type and order, wihtout regard to coumn names, which can be dangerous

`?data.table` contains : 

![image](https://user-images.githubusercontent.com/2779998/39952795-5540ac94-5553-11e8-9575-9630c59a038b.png)

Combining that with FAQ 2.14, does this answer your questions on how `X[Y]` works?

NB: Everyone, including me, is agreeing to making some changes.

ethanbsmith commented:
Hi @mattdowle,

Ive read the FAQ many times but it is quite dense an i clearly have not absorbed it all. I was clearly unaware of the "no on argument" behavior and and was assuming it was meant to work like an outer join. rather than you taking the time to explain to me what i got wrong, which is not productive for anyone, let me withdraw my current post and rethink it in terms of 2 discrete behaviors. I will resubmit my concerns after reexamining.

-E

ethanbsmith commented:
Please consider these examples: 

```
A = data.table(a = 1:5)
B = data.table(b = 2:6)
> A
   a
1: 1
2: 2
3: 3
4: 4
5: 5

> B
   b
1: 2
2: 3
3: 4
4: 5
5: 6

> A[B, on = .(a == b)] # Didn't really tell me anything
   a
1: 2
2: 3
3: 4
4: 5
5: 6

A2 = data.table(a = -7:0) #new data to be subset
> A2
    a
1: -7
2: -6
3: -5
4: -4
5: -3
6: -2
7: -1
8:  0

> A2[B, on = .(a == b)] # no different from the first query, even though A2 is completely different from A
   a
1: 2
2: 3
3: 4
4: 5
5: 6
```
We see that under current processing, information is being lost in 2 ways:
    1. the default output column name is misleading, as it implies the output value has come from A, when it really came from B. (The origin of most of the issues that lead to this PR)
    2. the result of the matching is not apparent from the output. One has to deeply understand the processing to make sense of the output

since under existing processing, we are returning only 1 column from a 2 column join, there is only 1 scenario where information cannot be lost: that is if the join columns have identical names AND values

We can extrapolate these examples in many ways but based on my reading of the existing issue descriptions, there is no problem with the number of rows returned in any situation (that has not been addressed or acknowledged as a bug).  So, i dont see much value in working out all the permutations of this. What i do think is important is that these 2 issues above apply to all forms of joins (inner, outer, rolling) and I would suggest that the solution being contemplated apply equally to all forms of joins

i do fully understand that these problems could be solved by use of %in% but that has 2 issues:
1. %in% only works on single column keys. if you have compound (multi-column) keys, you must use a join based solution
2. use of %in% requires that you understand enough about data.tables in the first place to know that the join model would fail in the first place 

now onto the solution:

```
> A2[B, on = .(a == b), .(x.a, i.b)] #explicitly including all joined columns avoids data loss
   x.a i.b
1:  NA   2
2:  NA   3
3:  NA   4
4:  NA   5
5:  NA   6

> A[B, on = .(a == b), .(x.a, i.b)] #same for the original query as we see in row 5
   x.a i.b
1:   2   2
2:   3   3
3:   4   4
4:   5   5
5:  NA   6

```
So, it would seem that unless you are dealing with an inner join where both tables have the same key column names, you are going to run into either problem 1, 2, or both.

I ask that once you are altering the output format of data.table join, you consider only solutions formats that addresses both of these problems. i.e.: _solutions that include all joined columns for all join forms._ 

I do see (based on prior discussions) that output for natural inner joins (where the values and column names are identical) can be optimized, and if this is desired, see no issue with adding this optimization for that subset. but to reiterate, the subset for which this works is only inner joins where the column names are identical. i do think that non-identical column names and/or anything other than an `inner-equi-join` will still suffer one of the problems above. 

The name and order of the columns does not seem too much of an issue, though probably the more uniform it is across join types, the less room for error. maybe treat the `on` parameter columns the way the `by` parameter columns are treated. i.e. not optional in the output (due credit: `data.table` handling/syntax of `by` is **WAY** better than SQL)

mattdowle commented:
Ethan, thanks for the last comment in parenthesis. It helps a lot to mention good things too.
I follow your points on these examples, but being 1-column examples makes me anxious. Nobody uses data.table like that in the real-world.  data.table thinking relies on there being some non-join columns (the purpose of doing the join). It's the contents of the non-join columns that make the result clearer.  I'm not saying no to changes, but I'm saying that if we are going to make breaking changes, then we should make those on real world examples, not 1-column abstract ones. Can you make more realistic examples with more than 1 column?

ethanbsmith commented:
```
DT <- data.table(key1 = 1:5, key2 = 2:6, key3 = 3:7, val1 = letters[1:5], val2 = letters[2:6], val3 = letters[3:7])
DT2 <- rbind(DT, list(1, 2, 3, NA_character_, NA_character_, NA_character_))
> DT
   key1 key2 key3 val1 val2 val3
1:    1    2    3    a    b    c
2:    2    3    4    b    c    d
3:    3    4    5    c    d    e
4:    4    5    6    d    e    f
5:    5    6    7    e    f    g
> DT2
   key1 key2 key3 val1 val2 val3
1:    1    2    3    a    b    c
2:    2    3    4    b    c    d
3:    3    4    5    c    d    e
4:    4    5    6    d    e    f
5:    5    6    7    e    f    g
6:    1    2    3 <NA> <NA> <NA>

#scenario 1
> DT[DT, on = c("key1==key1", "key2==key2", "key3==key3")]
   key1 key2 key3 val1 val2 val3 i.val1 i.val2 i.val3
1:    1    2    3    a    b    c      a      b      c
2:    2    3    4    b    c    d      b      c      d
3:    3    4    5    c    d    e      c      d      e
4:    4    5    6    d    e    f      d      e      f
5:    5    6    7    e    f    g      e      f      g
#here we see that i column names have i. prefix, as expected, key1,2,3 have no prefixes and we dont know which table side they came from. so we see column have name munging, but i get the argument that the values are the same so it doesn't matter

#scenario 1a
> DT2[DT2, on = c("key1==key1", "key2==key2", "key3==key3")]
   key1 key2 key3 val1 val2 val3 i.val1 i.val2 i.val3
1:    1    2    3    a    b    c      a      b      c
2:    1    2    3 <NA> <NA> <NA>      a      b      c
3:    2    3    4    b    c    d      b      c      d
4:    3    4    5    c    d    e      c      d      e
5:    4    5    6    d    e    f      d      e      f
6:    5    6    7    e    f    g      e      f      g
7:    1    2    3    a    b    c   <NA>   <NA>   <NA>
8:    1    2    3 <NA> <NA> <NA>   <NA>   <NA>   <NA>
#again, pretty much as expected

#scenario 2
> DT[DT, on = c("key1==key1", "key2==key1", "key3==key1")]
   key1 key2 key3 val1 val2 val3 i.key2 i.key3 i.val1 i.val2 i.val3
1:    1    1    1 <NA> <NA> <NA>      2      3      a      b      c
2:    2    2    2 <NA> <NA> <NA>      3      4      b      c      d
3:    3    3    3 <NA> <NA> <NA>      4      5      c      d      e
4:    4    4    4 <NA> <NA> <NA>      5      6      d      e      f
5:    5    5    5 <NA> <NA> <NA>      6      7      e      f      g
# where do the values of key1, key2, key3 come from? they are not present in either `x` nor `i`
# new columns in the output? that was unexpected. how do i know when im going to get those? 

#Scenario 2a
> DT2[DT2, on = c("key1==key1", "key2==key1", "key3==key1")]
   key1 key2 key3 val1 val2 val3 i.key2 i.key3 i.val1 i.val2 i.val3
1:    1    1    1 <NA> <NA> <NA>      2      3      a      b      c
2:    2    2    2 <NA> <NA> <NA>      3      4      b      c      d
3:    3    3    3 <NA> <NA> <NA>      4      5      c      d      e
4:    4    4    4 <NA> <NA> <NA>      5      6      d      e      f
5:    5    5    5 <NA> <NA> <NA>      6      7      e      f      g
6:    1    1    1 <NA> <NA> <NA>      2      3   <NA>   <NA>   <NA>
# expected base on scenario 2 output

#Scenario 2b
> DT2[DT2, on = c("key1==key1", "key1==key2", "key1==key3")]
   key1 key2 key3 val1 val2 val3 key1.1 key1.2 i.val1 i.val2 i.val3
1:    1   NA   NA <NA> <NA> <NA>      2      3      a      b      c
2:    2   NA   NA <NA> <NA> <NA>      3      4      b      c      d
3:    3   NA   NA <NA> <NA> <NA>      4      5      c      d      e
4:    4   NA   NA <NA> <NA> <NA>      5      6      d      e      f
5:    5   NA   NA <NA> <NA> <NA>      6      7      e      f      g
6:    1   NA   NA <NA> <NA> <NA>      2      3   <NA>   <NA>   <NA>
# so different from 2a. instead of i.key1,2,3 we have key1.1, key1.2, along with a bunch of NAs in key2,key3

```
I suspect that you will find these examples no more compelling than the single column ones. my point here is not to be tedious in these examples, but to show that a more complex example is not possible, since the nature of the issue will only ever be materialized in the output of the `on` columns, regardless of how many columns are involved. further, my point is to try show how deeply one needs to understand what is going on in the internals of how data.table implements these operations in order to anticipate the output format

as someone coming at this with years of relational database experience, i have found getting my head around some of this to be rather frustrating. I tend to do lots of prototyping, playing around with switching between inner, outer, equi and no - equi joins. I would really like the output format of the `on` columns to remain consistent across all that. All I really want to be able anticipate and trust to trust the output of each step without having to think about this ever again, as i have with relational databases for years.

However, I also completely understand the desire not to break existing code and to support `R` subsetting semantics! Ive been on the other side of that and I really, really do.

maybe a different way to think about this is not whats right vs wrong with only 1 solution, but more about perspective. What might be interesting is to give people optionality that suits their perspective: do you want output of `on` columns to work like r subsetting or like relational joins
i.e. Instead of changing the output formats for all subset of operations, maybe add something like a new "join-perspective=relational" option to have consistent number and names of columns from the `on` parameter in the output, across all join types and scenarios. Maybe even leave the existing R lookup semantics as the defaults for everything

sritchie73 commented:
Just wanted to drop a note here to say this is still on my radar - I'm been travelling / moving internationally so haven't had time to sit down and read all the comments. Hopefully should have some time in the next few weeks to revisit.

jangorecki commented:
@sritchie73 looking forward for your comments to recent posts here. We should consider to put it into 1.11.6 if whole new code is carefully isolated to be used only as option and previous behavior is retained.

mattdowle commented:
I ran all 684 CRAN and Bioconductor revdeps against this branch with `bothkeycols=TRUE`.  Only 4 packages failed:  CRAN (constellation, grattan, mtconnectR); Bioc (GenVisR).
Due to the nature of the change (data changing in a column of result) I think we'll have to be more careful by ensuring that every single usage of rolling join or non-equi-join is checked explicitly by the user in their code, not just rely on tests. There needs to be a way that they can find and check every single usage.
However, it's very good that at least we don't have a very large list of revdep breakage to deal with! This makes me more willing than I already was to make the breaking change.

mattdowle commented:
Future commits to the new branch in main please: `sritchie73-non-equi-key` rather than the remote one `sritchie73:non-equi-key`.

mattdowle commented:
Continues in main branch PR #3093 `sritchie73-non-equi-key` since I couldn't push to the remote branch `sritchie73:non-equi-key`.  Scott is now project member so can create branches directly in main project going forward.
